spec/lib/discourse_auth_cookie_spec.rb

```diff
@@ -0,0 +1,246 @@
+      expect(parsed.timestamp).to eq(timestamp.to_i)
+      expect(parsed.valid_for).to eq(valid_for.to_i)
+    end
+
+    it "does not require cookie fields to be in any particular order" do
+      token = SecureRandom.hex
+      timestamp = 4.hours.ago
+      valid_for = 100.days
+      cookie = DiscourseAuthCookie.new(
+        token: token,
+        user_id: user.id,
+        trust_level: user.trust_level,
+        timestamp: timestamp,
+        valid_for: valid_for
+      ).to_text
+
+      data = DiscourseAuthCookie::Encryptor.new.decrypt_and_verify(cookie)
+      data = data.split(",").shuffle.join(",")
+      cookie = DiscourseAuthCookie::Encryptor.new.encrypt_and_sign(data)
+
+      parsed = DiscourseAuthCookie.parse(cookie)
+      expect(parsed.token).to eq(token)
+      expect(parsed.user_id).to eq(user.id)
+      expect(parsed.trust_level).to eq(user.trust_level)
+      expect(parsed.timestamp).to eq(timestamp.to_i)
+      expect(parsed.valid_for).to eq(valid_for.to_i)
+    end
+  end
+
+  describe "#to_text" do
+    it "converts the instance to an encrypted string and signature" do
+      token = SecureRandom.hex
+      timestamp = 4.hours.ago
+      valid_for = 100.days
+      cookie = DiscourseAuthCookie.new(
+        token: token,
+        user_id: user.id,
+        trust_level: user.trust_level,
+        timestamp: timestamp,
+        valid_for: valid_for
+      ).to_text
+
+      data = DiscourseAuthCookie::Encryptor.new.decrypt_and_verify(cookie)
+      parts = data.split(",")
+      expect(parts.size).to eq(5)
+      expect(parts).to include("token:#{token}")
+      expect(parts).to include("id:#{user.id}")
+      expect(parts).to include("tl:#{user.trust_level}")
+      expect(parts).to include("time:#{timestamp.to_i}")
+      expect(parts).to include("valid:#{valid_for.to_i}")
+    end
+
+    it "works when some fields are nil" do
+      token = SecureRandom.hex
+      timestamp = 4.hours.ago
+      cookie = DiscourseAuthCookie.new(
+        token: token,
+        user_id: nil,
+        trust_level: nil,
+        timestamp: timestamp,
+        valid_for: nil
+      ).to_text
+
+      data = DiscourseAuthCookie::Encryptor.new.decrypt_and_verify(cookie)
+      parts = data.split(",")
+      expect(parts.size).to eq(5)
+      expect(parts).to include("token:#{token}")
+      expect(parts).to include("id:")
+      expect(parts).to include("tl:")
+      expect(parts).to include("time:#{timestamp.to_i}")
+      expect(parts).to include("valid:")
+    end
+  end
+
+  describe "#validate!" do
+    it "ensures token has the right length" do
+      token = SecureRandom.hex(8)
+      timestamp = 4.hours.ago
+      valid_for = 100.days
+      cookie = DiscourseAuthCookie.new(
+        token: token,
+        user_id: user.id,
+        trust_level: user.trust_level,
+        timestamp: timestamp,
+        valid_for: valid_for
+      )
+
+      expect {
+        cookie.validate!
+      }.to raise_error(DiscourseAuthCookie::InvalidCookie)
```

Comment:
Just out of curiosity, is it important for us to raise a more detailed error here?

Reply:
Do you mean like include what's specifically wrong with the cookie, e.g. too short or too old? We'd need/use the details anywhere if we were to add them, so I'd prefer to keep things simple for now.

Reply:
Yea I was wondering if we need a `DiscourseAuthCookie::TooShortCookie` error or something. If we don't need it, it is fine too was just curious :grin: